hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/UpgradeTask.java

```diff
@@ -0,0 +1,391 @@
+import org.apache.hadoop.ozone.container.keyvalue.KeyValueContainerData;
+import org.apache.hadoop.ozone.container.keyvalue.helpers.BlockUtils;
+import org.apache.hadoop.ozone.container.keyvalue.helpers.KeyValueContainerUtil;
+import org.apache.hadoop.ozone.container.metadata.DatanodeSchemaThreeDBDefinition;
+import org.apache.hadoop.ozone.container.metadata.DatanodeSchemaTwoDBDefinition;
+import org.apache.hadoop.ozone.container.metadata.DatanodeStore;
+import org.apache.hadoop.ozone.container.metadata.DatanodeStoreSchemaThreeImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * This class process upgrade v2 to v3 container.
+ */
+public class UpgradeTask {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(UpgradeTask.class);
+
+  private final ConfigurationSource config;
+  private final HddsVolume hddsVolume;
+  private final DatanodeStoreSchemaThreeImpl datanodeStoreSchemaThree;
+
+  private static final String BACKUP_CONTAINER_DATA_FILE_SUFFIX =
+      ".v2.container";
+
+  public UpgradeTask(ConfigurationSource config,
+                     HddsVolume hddsVolume,
+                     DatanodeStoreSchemaThreeImpl datanodeStoreSchemaThree) {
+    this.config = config;
+    this.hddsVolume = hddsVolume;
+    this.datanodeStoreSchemaThree = datanodeStoreSchemaThree;
+  }
+
+  public CompletableFuture<UpgradeManager.Result> getUpgradeFutureByVolume() {
+    return CompletableFuture.supplyAsync(() -> {
+
+      final UpgradeManager.Result result =
+          new UpgradeManager.Result(hddsVolume);
+
+      List<Result> resultList = new ArrayList<>();
+      final File hddsVolumeRootDir = hddsVolume.getHddsRootDir();
+
+      Preconditions.checkNotNull(hddsVolumeRootDir, "hddsVolumeRootDir" +
+          "cannot be null");
+
+      //filtering storage directory
+      File[] storageDirs = hddsVolumeRootDir.listFiles(File::isDirectory);
+
+      if (storageDirs == null) {
+        LOG.error("IO error for the volume {}, skipped loading",
+            hddsVolumeRootDir);
+        return null;
+      }
+
+
+      if (storageDirs.length > 0) {
+        File clusterIDDir = new File(hddsVolumeRootDir,
+            hddsVolume.getClusterID());
+        File idDir = clusterIDDir;
+        if (storageDirs.length == 1 && !clusterIDDir.exists()) {
+          idDir = storageDirs[0];
+        } else {
+          if (!clusterIDDir.exists()) {
+            LOG.error("Volume {} is in an inconsistent state. Expected " +
+                    "clusterID directory {} not found.", hddsVolumeRootDir,
+                clusterIDDir);
+            return null;
+          }
+        }
+
+        LOG.info("Start to upgrade containers on volume {}", hddsVolumeRootDir);
+        File currentDir = new File(idDir, Storage.STORAGE_DIR_CURRENT);
```

Comment:
There is one case, that for a fresh DN without any data written, there is no current directory.